### PR TITLE
[PORT] The roundstart SM Crystal delamination now spawns anomalies across the station + New Anomalies + Anomaly Code improvements

### PR DESCRIPTION
**Pull request by mystery3525**
## About The Pull Request

Ports:
- [x] [tgstation#55653](https://github.com/tgstation/tgstation/pull/55653)
- [x] [tgstation#56408](https://github.com/tgstation/tgstation/pull/56408)
- [x] [tgstation#63586](https://github.com/tgstation/tgstation/pull/63586)
- [x] [tgstation#65959](https://github.com/tgstation/tgstation/pull/65959)
- [x] [tgstation#66322](https://github.com/tgstation/tgstation/pull/66322)
- [x] [tgstation#66390](https://github.com/tgstation/tgstation/pull/66390)
- [x] [tgstation#66392](https://github.com/tgstation/tgstation/pull/66392)
- [x] [tgstation#66420](https://github.com/tgstation/tgstation/pull/66420)
- [x] [tgstation#66752](https://github.com/tgstation/tgstation/pull/66752)

## Refactors
- We now use `TRAIT_MADNESS_IMMUNE` when handling if a mob is affected by hallucinations.
- Supermatter delamination now lives inside of `supermatter_delamination.dm` with refactored code improvements
- Internally improved reactive armor code
- Reactive teleport armor now has an emp effect.

## Anomalies
- The **roundstart** Supermatter **Crystal** blowing up that doesn't produce a singo or tesla will produce anomalies across the station based on the explosion power. Half of the anomalies spawn immediately after explosion, and half spawn spread through 10 minutes after explosion, so be aware.
- New Hallucination anomaly and companion armor. Causes powerful supermatter hallucinations when viewed without protection.
- New Delimber anomaly and companion armor. Swaps random body parts throughout one's body.

## Why It's Good For The Game

> Makes the engine failure more than just a hole in your department and will have lasting repercussions on the round

> More non grief anomalies to add to the pool, expanding the amount of interesting anomalies that can spawn around the station

> Having the meson glasses having a typepath specific effect is gross, so do it with a trait instead. Then we
can make other goggles supermatter proof.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

image

</details>

## Changelog
:cl: coiax, Bazelart at TGI Friday's, TheBonded, Ghilker ported by mystery3525
add: Supermatter delamination now spawns anomalies across the station, both instantly and overtime.
refactor: Moves supermatter delamination to its own datum.
add: New hallucination anomaly which creates powerful hallucination pulses.
add: New delimber anomaly which randomizes and even rips off organs/limbs.
balance: Each of the reactive armors now have unique negative effects when emp'ed.
fix: Psychologists are no longer mistakenly told they have headaches
when near the supermatter.
/:cl:

### CODE REVIEW

**Comment by mystery3525** [comment]
Values can totally be adjusted for balance, or we could maybe make the # of anomalies based on population size to not be as punishing towards lowpop SM fuck-ups.

**Comment by Rukofamicom** [comment]
I love this and have questions and requests:
* Do zappy boi anomalies still not explode (Good if not, they explode very powerfully and wreak enough havoc just by zapping crew). 
* Can this please replace Tesla and Singularity since two of the three are generally an unfun way to end a round? The ratio of different types of anomalies can shift based on how the delamination is going instead. 
* Can the explosion be capped to a much smaller value and/or replaced with a stationwide weak (so IPCs can survive it) EMP? This will cause immediate malfunctions and setbacks that I think you're looking for without being over the top. 
* Are you willing to remove pyroclastic anomalies from the potentials and instead include other types of anomalies such as bluespace and wormholes? Even localized plasma flooding is extremely punishing, especially if they hit somewhere like a main hall.

**Comment by mystery3525** [comment]
I believe /tg/ is working on adding different anomaly types to spice things up, though I could expand the scope of this PR to their general consequences of the SM delaminating, I need to work on fixing the bugs found in testing.

image

EDIT: How do I set this as a draft, Here I'll just put a [DNM]

**Comment by Rukofamicom** [comment]
image

See the blue text? Find it and click it to convert into a draft.

**Comment by Cartlord** [comment]
The fucking engine exploding is already a really big deal with serious repercussions to the round, I don't know what you're on about. Even ignoring the fat hole in Engineering, there's still the fact that _we no longer have power_ unless the Miners manage to haul up plasma sheets for a P.A.C.M.A.N. or someone actually knows how to operate the turbine **and** is actually able to get the damn thing set up before the lack of power kills all the pumps leading to it.

**Comment by Mat05usz** [comment]
> Values can totally be adjusted for balance,

I might be wrong, but I think just a regular delam makes 3/4 of the station a shithole riddled with holes and plasma because so many anomalies spawn. Values most certainly will have to be adjusted, unless you want every SM delam end in shuttle being called.

**Comment by mystery3525** [comment]
(Wasn't blue for me)

I recolonize the issues that were addressed on the /tg/ side of things, and I did apply the port to nerf the flux anomalies. I'll port the new anomalies once I can get the code to actually spawn anomalies (local testing has shown absolutely none have spawned). However, we are in a much more better position than their players in the fact that we have options for power after the dust has settled, solar panels as an aftermath, turbine on most maps since atmos won't be completely destroyed, even the ghetto antimatter in Cargo's Warehouse. The extra anomaly types will reduce the chances for the more disruptive pyro and blackhole anomalies from spawning.

**Comment by mystery3525** [comment]
Might as well port anomaly additions and reactive armor improvements.

**Comment by Rukofamicom** [comment]
I maintain that this shouldn't include pyroclastic or exploding energetic anomalies at all. There were already SM specific energetic anomalies that don't explode.

**Comment by MNarath1** [comment]
Now important distinction are we talking about actual delamination (integrity goes down) or just overloading the sm (disco sm over 5000 MeV) that spawn anomalies around the station

**Comment by mystery3525** [comment]
> Now important distinction are we talking about actual delamination (integrity goes down) or just overloading the sm (disco sm over 5000 MeV) that spawn anomalies around the station

The resultant explosion which qdel's the SM crystal causes the anomalies across the explosion.

**Comment by Howluinb** [comment]
I really like this, especially the delimber one.

**Comment by mystery3525** [comment]
Fully greenlit, ready for review.

File: code/__DEFINES/research/anomalies.dm
Comment (PowerfulBacon) [inline code/__DEFINES/research/anomalies.dm:14]: I feel like these should be closer together, there is a big difference between the top and lowest since the gravitational anomaly is 55x more likely to spawn than the vortex. Maybe something more like 60, 50, 35 and 20

File: code/modules/clothing/suits/reactive_armour.dm
Comment (PowerfulBacon) [inline code/modules/clothing/suits/reactive_armour.dm]: This message feels weird, it should make more sense

File: code/modules/clothing/suits/reactive_armour.dm
Comment (PowerfulBacon) [inline code/modules/clothing/suits/reactive_armour.dm:264]: Is this meant to always give the wearer TESLA_IGNORE_1 even if they didn't have it to begin with?

File: code/modules/clothing/suits/reactive_armour.dm
Comment (PowerfulBacon) [inline code/modules/clothing/suits/reactive_armour.dm]: Suggestion: spawn a glass table instead

File: code/modules/clothing/suits/reactive_armour.dm
Comment (PowerfulBacon) [inline code/modules/clothing/suits/reactive_armour.dm:365]: This will remove hallucinations from the owner if they are over 150.
It should be something like
```
owner.halluncation = min(owner.halluncation + 25, 150)
```

File: code/modules/clothing/suits/reactive_armour.dm
Comment (PowerfulBacon) [inline code/modules/clothing/suits/reactive_armour.dm]: this should definitely message admins and/or be logged

File: code/__DEFINES/research/anomalies.dm
Comment (mystery3525) [reply code/__DEFINES/research/anomalies.dm:14]: The gravitational and hallucination anomalies are by far the least consequential anomalies. The Pyro anomaly plasmafires an area if not secured, and the vortext anomaly straight exposes a large area to space.

File: code/modules/clothing/suits/reactive_armour.dm
Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm:264]: Yeah looks like it was adapted since /tg/ uses a different method for applying tesla ignores, but the reactive tesla armor sets TESLA_IGNORE_1 on the owner flags. This disables it to temporarily electrocute and re-enables it.

Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm:264]: When they put it on, they recieve TESLA_IGNORE_1. And when they take it off they get TESLA_IGNORE_1 removed, which is honestly shitcode since there is at least another way to get TESLA_IGNORE_1, but we'd need to port the /tg/ fixes for that.

File: code/modules/clothing/suits/reactive_armour.dm
Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm:365]: If they are over 150, they should be **clamped** to 150 since it is the highest possible? https://www.byond.com/docs/ref/#/proc/clamp

File: code/__DEFINES/research/anomalies.dm
Comment (mystery3525) [reply code/__DEFINES/research/anomalies.dm:14]: I would recommend waiting until a test merge and a couple explosions to modify these rates, as this is the amount /tg/ uses after a few nerf PR's.

File: code/modules/clothing/suits/reactive_armour.dm
Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm]: Used same logging method as voice of god

File: code/modules/clothing/suits/reactive_armour.dm
Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm]: and also resorted to more proven do_teleport (same as reactive teleport)

File: code/game/objects/effects/anomalies.dm
Comment (Sarchutar) [inline code/game/objects/effects/anomalies.dm:490]: just saying, this literally has the ability to completely remove you if it chooses any brain, it will literally qdel you when `Insert()` calls `Remove()` (which puts you in your brain and then deletes it)

File: code/modules/clothing/suits/reactive_armour.dm
Comment (Sarchutar) [inline code/modules/clothing/suits/reactive_armour.dm]:
```suggestion
	addtimer(CALLBACK(src, .proc/end_stealth, owner), stealth_time)
```
this may belong here (stealth_time is currently unused otherwise)

File: code/modules/clothing/suits/reactive_armour.dm
Comment (Sarchutar) [inline code/modules/clothing/suits/reactive_armour.dm:195]: this will actually never trigger since `reactivearmor_cooldown_duration` is 0 (default from parent)

the only reactive armor who actually has that var set is the teleport one (which also makes it the only one with an actual cooldown, the other ones can always be triggered because of this)

File: code/modules/clothing/suits/reactive_armour.dm
Comment (Sarchutar) [inline code/modules/clothing/suits/reactive_armour.dm]: this only creates a decoy, the stealth/invisible mechanic is not in here anymore
`owner.alpha = 0`

File: code/modules/clothing/suits/reactive_armour.dm
Comment (Sarchutar) [inline code/modules/clothing/suits/reactive_armour.dm]: these seem to be unused

File: code/__DEFINES/traits.dm
Comment (Sarchutar) [inline code/__DEFINES/traits.dm]: this trait doesn't actually do what it says it does right now (right now it's the same as `TRAIT_MADNESS_IMMUNE` and is given to the Psychiatrist which gets both anyway)

**Comment by PowerfulBacon** [comment]
Unaddressed reviews from Sarchutar

File: code/modules/clothing/suits/reactive_armour.dm
Comment (mystery3525) [reply code/modules/clothing/suits/reactive_armour.dm:195]: default cooldown for all is now 5 seconds

File: code/game/objects/effects/anomalies.dm
Comment (mystery3525) [reply code/game/objects/effects/anomalies.dm:490]: blacklisted brains, xenoegg spawners and romerol tumors

**Comment by mystery3525** [comment]
yeah

**Review by PowerfulBacon — APPROVED** [review]

**Pull request merged**
